app/notifiers/notifier/base.rb

```diff
@@ -0,0 +1,84 @@
+# The base class of notifiers. This is meant to be called by the Notifications Framework
+#
+# @api notifications
+class Notifier::Base
+  include ApplicationNotificationsConcern
+
+  class ActivityWrapper < SimpleDelegator
+    # Send notifications according to input type and recipient
+    #
+    # @param [Object] recipient The recipient of the notification
+    # @param [Symbol] type The type of notification
+    def notify(recipient, type)
+      super(recipient, type).tap do |notification|
+        case recipient
+        when Course
+          @pending_emails += @notifier.send(:email_course, notification) if type == :email
+        when User
+          @pending_emails << @notifier.send(:email_user, notification) if type == :email
+        end
+      end
+
+      self
+    end
+
+    # Save activity and send out pending emails
+    def save(*)
+      super.tap do |result|
+        @pending_emails.each(&:deliver_later) if result
```

Comment:
what do you intend to do after delivering pending_emails? I think the delivered emails can be left alone.